File: app/src/main/java/com/vladiknt/giph/RegisterActivity.java
package com.vladiknt.giph;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.icu.util.DateInterval;
import android.icu.util.GregorianCalendar;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public class RegisterActivity extends AppCompatActivity {

    private FirebaseAuth mAuth;
    private FirebaseFirestore db;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);
        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();
    }

    public void registerUser(View view) {
        EditText et = findViewById(R.id.regMail);
        String email = et.getText().toString();
        et = findViewById(R.id.regPass1);
        String pass1 = et.getText().toString();
        et = findViewById(R.id.regPass2);
        String pass2 = et.getText().toString();
        et = findViewById(R.id.regNickname);
        String nickname = et.getText().toString();

        // Проверка на возраст
        et = findViewById(R.id.regAge);
        try {
            int age = Integer.parseInt(et.getText().toString());
            if (age < 18) {
                Toast.makeText(RegisterActivity.this, "Our app is only for adults.", Toast.LENGTH_SHORT).show();
                return;
            }
        } catch (Exception e) {
            Toast.makeText(RegisterActivity.this, "Enter correct age (single number).", Toast.LENGTH_SHORT).show();
            return;
        }

        // Подготовка файлов пользователя для добавления в БД
        Map<String, Object> userLevels = new HashMap<>();
        userLevels.put("hentai", 0);
        userLevels.put("asians", 0);
        userLevels.put("coins", 10);
        SimpleDateFormat formatter = new SimpleDateFormat("dd.MM.yyyy", Locale.UK);
        String dateTime = formatter.format(new Date());
        Map<String, Object> userInfo = new HashMap<>();
        userInfo.put("nickname", nickname);
        userInfo.put("dateJoined", dateTime);
        // Регистрация
        CheckBox license = findViewById(R.id.createAccLicense);
        if (license.isChecked()) {
            if (pass1.equals(pass2)) {
                mAuth.createUserWithEmailAndPassword(email, pass1)
                        .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                            @Override
                            public void onComplete(@NonNull Task<AuthResult> task) {
                                if (task.isSuccessful()) {
                                    db.collection("users").document(mAuth.getCurrentUser().getUid()).set(userInfo).addOnCompleteListener(task1 -> {
                                        if (task1.isSuccessful()) {
                                            db.collection("levels").document(mAuth.getCurrentUser().getUid()).set(userLevels).addOnCompleteListener(task2 -> {
                                                if (task2.isSuccessful()) {
                                                    mAuth.getCurrentUser().sendEmailVerification();
                                                    Toast.makeText(RegisterActivity.this, "Please, activate your profile in letter.", Toast.LENGTH_SHORT).show();
                                                    finish();
                                                } else
                                                    Toast.makeText(RegisterActivity.this, "Error while adding user levels.", Toast.LENGTH_SHORT).show();
                                            });
                                        } else
                                            Toast.makeText(RegisterActivity.this, "Error while adding user info.", Toast.LENGTH_SHORT).show();
                                    });
                                } else
                                    Toast.makeText(RegisterActivity.this, "Authentication failed.", Toast.LENGTH_SHORT).show();
                            }
                        });
            } else {
                et = findViewById(R.id.regPass1);
                et.clearComposingText();
                et = findViewById(R.id.regPass2);
                et.clearComposingText();
                Toast.makeText(RegisterActivity.this, "Enter your passwords again.", Toast.LENGTH_SHORT).show();
            }
        } else
            Toast.makeText(RegisterActivity.this, "Please, agree with license.", Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/vladiknt/giph/RegisterActivity.kt
package com.vladiknt.giph

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.CheckBox
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.HashMap

class RegisterActivity : AppCompatActivity() {
    private var mAuth: FirebaseAuth? = null
    private var db: FirebaseFirestore? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        mAuth = FirebaseAuth.getInstance()
        db = FirebaseFirestore.getInstance()
    }

    fun goToLicense(view: View?) {
        val lic = Intent(this, LicenseActivity::class.java)
        startActivity(lic)
    }

    fun registerUser(view: View?) {
        var et = findViewById<EditText>(R.id.regMail)
        val email = et.text.toString()
        et = findViewById(R.id.regPass1)
        val pass1 = et.text.toString()
        et = findViewById(R.id.regPass2)
        val pass2 = et.text.toString()
        et = findViewById(R.id.regNickname)
        val nickname = et.text.toString()

        // Проверка на возраст
        et = findViewById(R.id.regAge)
        try {
            val age = et.text.toString().toInt()
            if (age < 18) {
                Toast.makeText(this, "Our app is only for adults.", Toast.LENGTH_SHORT).show()
                return
            }
        } catch (e: Exception) {
            Toast.makeText(this, "Enter correct age (single number).", Toast.LENGTH_SHORT).show()
            return
        }

        // Подготовка файлов пользователя для добавления в БД
        val userLevels: MutableMap<String, Any> = HashMap()
        userLevels["hentai"] = 0
        userLevels["asians"] = 0
        userLevels["coins"] = 10
        val formatter = SimpleDateFormat("dd.MM.yyyy", Locale.UK)
        val dateTime: String = formatter.format(Date())
        val userInfo: MutableMap<String, Any> = HashMap()
        userInfo["nickname"] = nickname
        userInfo["dateJoined"] = dateTime
        // Регистрация
        val license = findViewById<CheckBox>(R.id.createAccLicense)
        if (license.isChecked) {
            if (pass1 == pass2) {
                mAuth!!.createUserWithEmailAndPassword(email, pass1)
                    .addOnCompleteListener(this) { task ->
                        if (task.isSuccessful) {
                            db!!.collection("users").document(mAuth!!.currentUser!!.uid).set(userInfo)
                                .addOnCompleteListener { task1: Task<Void?> ->
                                    if (task1.isSuccessful) {
                                        db!!.collection("levels").document(mAuth!!.currentUser!!.uid).set(userLevels)
                                            .addOnCompleteListener { task2: Task<Void?> ->
                                                if (task2.isSuccessful) {
                                                    mAuth!!.currentUser!!.sendEmailVerification()
                                                    Toast.makeText(this, "Please, activate your profile in letter.", Toast.LENGTH_SHORT).show()
                                                    finish()
                                                } else Toast.makeText(this, "Error while adding user levels.", Toast.LENGTH_SHORT).show()
                                            }
                                    } else Toast.makeText(this, "Error while adding user info.", Toast.LENGTH_SHORT).show()
                                }
                        } else Toast.makeText(this, "Authentication failed.", Toast.LENGTH_SHORT).show()
                    }
            } else {
                et = findViewById(R.id.regPass1)
                et.clearComposingText()
                et = findViewById(R.id.regPass2)
                et.clearComposingText()
                Toast.makeText(this, "Enter your passwords again.", Toast.LENGTH_SHORT).show()
            }
        } else Toast.makeText(this, "Please, agree with license.", Toast.LENGTH_SHORT).show()
    }
}